### DISCUSSION TITLE
New Python Examples

### DISCUSSION THREAD

**Discussion by nathanlukeIDEA** [idea-statica/ideastatica-public#212] [Show and tell]
There is a new Python example available that explains step-by-step how to create a steel frame structure and connection using IDEA Open Model (IOM) with Python. 

Please check out the [using Python](https://github.com/idea-statica/ideastatica-public/wiki/Using-Python-with-IDEA-Packages) page on our wiki and scroll to the examples. 

Please let us know what else you would like to see more of.

**Comment by umesh14398** [comment]
Hi Nathan,

My name is Uma Mahesh.
I am working with IdeaStatica for one of the automation tasks and I have chosen Python for that. I am facing an issue in establishing the connection with IdeaStatica using Python. I followed the guidelines given in the some of the pages as shown below.
https://github.com/idea-statica/ideastatica-public/wiki/Using-Python-with-IDEA-Packages
https://github.com/stevenverwer/ideastaticapy/blob/main/ideastaticapy/ideaConnection.py#L37

Below is the code snippet which I am using it to connect with IdeaStatica but There seems to be a problem with a function(factory.Create()) at line 21. I am not sure what could be the issue here. I want to know your comments/suggestions to resolve this issue. Looking forward to your response
![image](https://github.com/idea-statica/ideastatica-public/assets/43907961/f5c8d44f-e148-4073-92e4-e09ad6a0c8a3)

**Reply by nathanlukeIDEA** [reply]
Hi @umesh14398, Your code is correct. There was a small bug in our Plugin where it would time out sometimes when creating the connection with the app. This only happens sometimes, therefore If you try and re-run your script it can solve the issue. However, I would recommend updating the connection app to the latest version and it should fix this occurrence.

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA 
Thank you so much for your response. 
I have updated the connection app to the latest version(ver 23.0.5) and tried to run the script a few times. I did it in Python and also in C# but faced the same issue in both cases. I am creating the new project as shown below in the IdeaStatic and running the script. I assume that is not a problem. Is there anything I can do to make the script work? 
<img width="964" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/43907961/d3fa57b4-6a5c-4853-8a87-fb812afa6824">

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA ,

Good Morning, 
Hope you are doing well,
Could you please look into my issue? I just got stuck at the beginning level of my task. 
I am not able to connect to IdeaStatica and still facing the same issue that i have explained above. 

Regards,
Uma Mahesh

**Reply by mpospisil** [reply]
Hi Uma,
Nathan is on holiday. Can you set the log level of ideastatica diagnostics to the debug level

rerun your script and send us the zip file of our log files. (See https://github.com/idea-statica/ideastatica-public/wiki/Guide-Diagnostics-in-Checkbot )

I would appreciate all files from 
C:\Users\USER NAME\AppData\Local\Temp\IdeaStatiCa\Logs\

I suppose it could be an issue related to the license. Check you have reserve seats :

![MicrosoftTeams-image (2)](https://github.com/idea-statica/ideastatica-public/assets/19650653/e85d29da-6c5b-42e0-b239-b569cbec5fe1)

**Reply by umesh14398** [reply]
Hi @mpospisil 
Thank you for your response. 
I set the log level to debug mode and ran the script. Please find the log files after running the script. Let me know your comments.

[IdeaStatiCa_Logfiles.zip](https://github.com/idea-statica/ideastatica-public/files/12927952/IdeaStatiCa_Logfiles.zip)

**Reply by mpospisil** [reply]
I can see in your log file _IdeaStatiCaConnHiddenCalculator.log_

2023-10-17 16:39:22.770 +05:30 [Debug] (32876-9) [license.seat.handlerbase] Seat reservation has already been done with features System.Collections.Generic.List`1[IdeaStatiCa.License.Seat.FeatureId]
2023-10-17 16:39:22.781 +05:30 [Debug] (32876-9) [license.seat.handlerbase] The application demand IdeaStatiCa.License.Seat.FeatureId[] is not satisfied with seat reserved features System.Collections.Generic.List`1[IdeaStatiCa.License.Seat.FeatureId].


check reserved seats on your PC as I wrote in my previous message

**Reply by umesh14398** [reply]
hi @mpospisil 
I didn't reserve any seats actually but i have access to the license
<img width="595" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/43907961/9c246201-075f-410f-82eb-d5c8c95e52c7">

**Reply by mpospisil** [reply]
There was a bug in ConnHiddenCalculator.exe. It was blocked in case of valid license for products  Steel, Complete or Enterprise
It is fixed. This change will be available in the v 23.1 Beta which will be published today
The official release of the version 23.1 will be the next week

**Reply by umesh14398** [reply]
Hi @mpospisil 
Thank you so much for the update. I am looking forward to the release of the version 23.1. I hope this issue will be resolved after updating the Idea Statica to the latest version.

**Reply by Balkmannen** [reply]
Excellent! I have been waiting for this. Can anyone access the beta, or will we have to wait for the official release?

**Reply by mpospisil** [reply]
you can download [ 23.1 beta  version](https://download.ideastatica.com/beta/idea-statica.msi)

**Reply by umesh14398** [reply]
Hi @mpospisil, Good Morning
When can we expect the official release of the version 23.1?

**Reply by mpospisil** [reply]
It is planenned for today

**Reply by umesh14398** [reply]
Hi @mpospisil 
Thank you so much for your response and support. 
I have a small question. I am trying to use the expressions to get the cross-section information of the beam or column in the developer tab. I was able to get the width and height of the beam but I could not get the exact dimensions of the flange thickness and web thickness. Am I doing anything wrong here to access the attributes? 
![image](https://github.com/idea-statica/ideastatica-public/assets/43907961/ce88900c-05e5-43aa-931d-afc57e15d29a)

**Reply by mpospisil** [reply]
Hi, Nathan will answer your you question.

**Reply by nathanlukeIDEA** [reply]
Hi @umesh14398,

You can use the following to get the plate thickness. :)

GetBeamPlateThickness('B', 'Web')
GetBeamPlateThickness('B', 'TopFlange')
GetBeamPlateThickness('B', 'BottomFlange')

Cheers

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA 
Good Morning, Thank you so much for your response. 
Can I get the link or source for API to implement functionalities using Python? Any API or API documentation will be very useful.

**Reply by nathanlukeIDEA** [reply]
hi @umesh14398,

link to some documentation. 
https://developer.ideastatica.com/api-plugin/IdeaStatiCa.Plugin.ConnectionHiddenCheckClient.html

It is still a work in progress.

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA 
Thank you so much.

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA 
One small question. I am trying to extract the properties of the beam and column.  I am unable to find a way to get the properties of the beam through code. I want to the properties like width, height, and thickness(flange and web) of the existing I section. Can you please help me out with how to get it? I am trying to execute it through Python

**Reply by nathanlukeIDEA** [reply]
Hi @umesh14398,

Currently we can't provide properties of members directly by code, but you can evaluate expressions. This is similar to the expressions that can evaluated in the Parameters interface, but through an API call. 

```python
height = client.EvaluateExpression(connection, "GetValue('COL', 'CrossSection.Bounds.Height')", None)
width = client.EvaluateExpression(connection, "GetValue('COL', 'CrossSection.Bounds.Width')", None)
```

![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/bef75fec-bc15-42b6-9f4d-32171b7e42ae)


Let me know if this helps.

Cheers
Nathan

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA 
Thank you so much for your response. it is very useful information.

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA , Thank you so much for your reply, Sorry if I am troubling you. 
I would like to know how to **GET** the **CROSS SECTIONS** of the members, ex: cross-section applied to beam or column,
also, I want to **SET** various **CROSS SECTIONS** to the beam and column through code.

**Reply by nathanlukeIDEA** [reply]
hi @umesh14398 here is a script that will allow you to get and set the member size. 
[Change Member Size_API.zip](https://github.com/idea-statica/ideastatica-public/files/13291340/Change.Member.Size_API.zip)

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA 
Thank you for the response. In the script it is given then that getting all the cross sections in a given project. I have few doubts
1. Any function to get the cross section for particular beam or cross but not all the cross sections
2. Can we set the parameters for the operations like EP1, SM1, STIFF1 etc.. As per the document in the forum, i observed that operation functions are only for internal use.

I want to perform the connection optimization actually. I have few cross sections(Ex: WPB 900x300x291.45, WPB 600x300x177.77)., I need to create the cross sections and If i set the particular cross section to beam then i need to update the  various parameters for the operations as well the cross section parameters'. I have done it in the developers tab but now i want to do the same through code as well. 
Can you please guid me in this regard.

**Reply by nathanlukeIDEA** [reply]
Hi @umesh14398 you can use this expression to get the mprl name: 
`GetValue('B','CrossSection.mprlName')` or this to get the actual name `GetValue('B','CrossSection.name')`

Anything available in the model properties dialogue can be set including cross-sections. If you want to set a particular cross-section to a member, you should add the ones you may want to use to the database in the file. Then simply set up a parameter that changes the cross-section.

![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/afad61bd-46bf-4ea6-8bec-14d0047c537d)


I assume you understand that you can update parameters using the API, but just incase:
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/6d438b58-78c5-4933-b711-b6ca2977af21)

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA  Thank You for the response. Your response gives clear information.  Yes, I understood how the parameters will be extracted and applied.

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA 
How can we assign the layer parameters in the developer tab and also through code? I have sent the parameters in the string format, The values are accepted but values are not getting updated in the model.
![image](https://github.com/idea-statica/ideastatica-public/assets/43907961/c7bf2c22-1eec-4ee4-94d3-5a762aabb7ff)
![image](https://github.com/idea-statica/ideastatica-public/assets/43907961/d00082cf-74f4-4870-9525-a16a7a34abd7)

I also tried to do the same through code, but still got the have the same issue. Can you please help me out with this issue?

**Reply by nathanlukeIDEA** [reply]
Hi @umesh14398. I don't believe the comma between -0.120 and -0.065 is a valid input right? it should just be a space? If you remove the comma it should work. Becuase we don't know exactly what the input is until we try and apply the string to the property we can't validate it here at the parameter level. But you should get an error message in the scene that does not seem to be appearing I will look into this.

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA , Yes, you are right. There should not be a comma but I added it by mistake. Now it is updating the model as well. 
After the applying the paramters, if I extract the parameters from the model, then i could see an error in the output window as shown below. Values are getting updated properly in the model i am not sure why i am getting the error. 
<img width="764" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/43907961/7aa60d65-4d46-4553-961f-7a45bb556530">

**Reply by nathanlukeIDEA** [reply]
Hi @umesh14398, this looks like a bug on our side. And just confirming this is not blocking you?

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA , No it not blocking me anything. I mean it's affecting in any way.

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA ,
Is it possible to create the new operation through code? I would like to create some operations and set the parameters. Is there any function to do that? I looked at the API documentation but I didn't find any functionality related to creating a new operation. I have created one template file but I could not apply the same template to other connections because of the topology issues. Can we make one template file for multiple joints..?  
![image](https://github.com/idea-statica/ideastatica-public/assets/43907961/e952ad58-920e-406e-8c55-44eb82664239)

**Reply by nathanlukeIDEA** [reply]
Hi @umesh14398, you cannot currently add an operation directly through the API. 

However, you can add 'simple' templates between a supporting member and a list of incoming members which theoretically can only hold one operation + any parameters you have predefined.

```python
client.ApplySimpleTemplate(connectionId, templateFilePath, connTemplateSettings, supportingMemberId, attachedMembersId's)

```

Let me know if this helps. We definitely could look at the ability to add operations through API. Could you please explain why you want to add the operation directly from API and what benefit this would have over using a template?

Cheers
Nathan

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA , Thank you for the response. The reason for creating a new operation through API is, that I have created one template file that contains all the operations but if I apply the same template to other connections containing various other joints then there is a topology issue. It is saying that the Current Template is created for Topology Y+; Z+; and cannot be applied to Y+;Y-;Z+;Z-.  It's not allowing one template for other joints. So that's why I am thinking of creating the operations through API. Now I wanted to do it using the ApplySimpleTemplate function but it is not working. 
![image](https://github.com/idea-statica/ideastatica-public/assets/43907961/e4f28215-3515-4858-a900-03e8a38ef0fb)

![image](https://github.com/idea-statica/ideastatica-public/assets/43907961/40b43eee-a4fa-4f9a-a23e-b889bdc9ae07)

**Reply by umesh14398** [reply]
Hi @nathanlukeIDEA , Can you look into this issue once ? I am not sure what is causing the problem here.

**Reply by nathanlukeIDEA** [reply]
I presume this is not working because you are not creating a .Net list type for the incoming members. pythonNet is not great at transferring .Net lists. If you replace [2] with attachedMember see if it works.

```python
attachedMembers = List[int]()
attachedMembers.Add(2)
```
